### PR TITLE
AdressBook Feature

### PR DESCRIPTION
**Pull request by Moetez16**

### CODE REVIEW

File: app/grails-app/views/templates/cpa/_personFormModal.gsp [hunk 1]
```diff
@@ -1,12 +1,15 @@
-<%@ page import="de.laser.PersonRole; de.laser.Contact; de.laser.Person; de.laser.FormService; de.laser.helper.RDStore; de.laser.RefdataValue;de.laser.RefdataCategory;de.laser.helper.RDConstants" %>
+<%@ page import="de.laser.properties.PropertyDefinition; com.k_int.kbplus.PersonRole; de.laser.Contact; de.laser.Person; de.laser.FormService; de.laser.helper.RDStore; de.laser.RefdataValue;de.laser.RefdataCategory;de.laser.helper.RDConstants" %>
```
Comment (klodav) [inline app/grails-app/views/templates/cpa/_personFormModal.gsp:1]: de.laser.PersonRole

File: app/grails-app/views/templates/cpa/_personFormModal.gsp [hunk 2]
```diff
@@ -193,7 +223,28 @@
...
+                    </div>
+
+
+                    <div class="field twelve wide fieldcontain">
+                        <label></label>
+                        <g:textField id="content" name="content" value="${contactInstance?.content}"/>
+                    </div>
+                </div>
+
+                <div id="contcatElements"></div>
```
Comment (klodav) [inline app/grails-app/views/templates/cpa/_personFormModal.gsp:247]: contcatElements or contactElements?

File: app/grails-app/views/templates/cpa/_personFormModal.gsp [hunk 3]
```diff
@@ -251,26 +334,32 @@
                 }
             });
 
-            var elementCount = 0;
+            var contactElementCount = 0;
+            var addressElementCount = 0;
+
+            var contcatContainer = $(document.createElement('div'));
+            $(contcatContainer).attr('id', 'contcatElementsContainer');
```
Comment (klodav) [inline app/grails-app/views/templates/cpa/_personFormModal.gsp:341]: contcatElementsContainer or contactElementsContainer?

File: app/grails-app/views/templates/cpa/_personFormModal.gsp [hunk 4]
```diff
@@ -251,26 +334,32 @@
...
+                            $(contcatContainer).append(data);
+                            $('#contactFields').attr('id', 'contactFields' + contactElementCount);
 
-                            $('#divElements').after(container);
+                            $('#contcatElements').after(contcatContainer);
                         } else {
                             $('#addContactElement').attr('class', 'ui icon button disable');
                             $('#addContactElement').attr('disabled', 'disabled');
                         }
+                        r2d2.initDynamicSemuiStuff('#contcatElementsContainer');
```
Comment (klodav) [inline app/grails-app/views/templates/cpa/_personFormModal.gsp:362]: contcatElementsContainer or contactElementsContainer?

File: app/grails-app/views/templates/cpa/_personFormModal.gsp [hunk 5]
```diff
@@ -251,26 +334,32 @@
...
 
-                            elementCount = elementCount + 1;
-                            $(container).append(data);
-                            $('#contactFields').attr('id', 'contactFields' + elementCount);
+                            contactElementCount = contactElementCount + 1;
+                            $(contcatContainer).append(data);
+                            $('#contactFields').attr('id', 'contactFields' + contactElementCount);
 
-                            $('#divElements').after(container);
+                            $('#contcatElements').after(contcatContainer);
```
Comment (klodav) [inline app/grails-app/views/templates/cpa/_personFormModal.gsp:357]: contcat or contact?

File: app/grails-app/views/myInstitution/addressbook.gsp [hunk 6]
```diff
@@ -162,4 +123,30 @@ import="de.laser.PersonRole; de.laser.helper.RDStore; de.laser.helper.RDConstant
                 total="${num_visiblePersons}"/>
 
 </body>
+
+
+<g:javascript>
+    function personCreate(contactFor) {
+        var url = '<g:createLink controller="ajax" action="createPerson"/>?contactFor='+contactFor+'&showAddresses='+${false}+'&showContacts='+${true};
```
Comment (klodav) [inline app/grails-app/views/myInstitution/addressbook.gsp:130]: controller="ajaxHtml"

Comment (klodav) [reply app/grails-app/views/myInstitution/addressbook.gsp:130]: '&showAddresses=false&showContacts=true'

File: app/grails-app/views/myInstitution/myPublicContacts.gsp [hunk 7]
```diff
@@ -242,5 +225,53 @@
     $(document).ready(function () {
         $('.tabular.menu .item').tab()
     });
+
+    function personCreate(contactFor) {
+        var url = '<g:createLink controller="ajax" action="createPerson"/>?contactFor='+contactFor+'&showAddresses='+${false}+'&showContacts='+${true};
```
Comment (klodav) [inline app/grails-app/views/myInstitution/myPublicContacts.gsp:230]: controller="ajaxHtml"

Comment (klodav) [reply app/grails-app/views/myInstitution/myPublicContacts.gsp:230]: '&showAddresses=false&showContacts=true'

File: app/grails-app/views/myInstitution/myPublicContacts.gsp [hunk 8]
```diff
@@ -242,5 +225,53 @@
...
+                        r2d2.initDynamicSemuiStuff('#personModal');
+                        r2d2.initDynamicXEditableStuff('#personModal');
+                    }
+                }).modal('show');
+            }
+        });
+    }
+
+    function addresscreate_org(orgId) {
+            var url = '<g:createLink controller="ajax" action="createAddress"/>'+'?orgId='+orgId;
```
Comment (klodav) [inline app/grails-app/views/myInstitution/myPublicContacts.gsp:252]: controller="ajaxHtml"

File: app/grails-app/views/templates/cpa/_address_table.gsp [hunk 9]
```diff
@@ -25,7 +25,13 @@
                 ${c + 1}
             </td>
             <td>
-                ${address.type?.getI10n('value')}
+                <div class="ui divided middle aligned list la-flex-list ">
+                <g:each in="${address.type.sort{it?.getI10n('value')}}" var="type">
+                    <div class="ui item ">
+                        ${type?.getI10n('value')}
```
Comment (klodav) [inline app/grails-app/views/templates/cpa/_address_table.gsp:31]: type?

File: app/grails-app/views/templates/_copyFilteredEmailAddresses.gsp [hunk 10]
```diff
@@ -4,7 +4,7 @@
 
 <g:set var="modalID" value="${modalID ?: 'copyFilteredEmailAddresses_ajaxModal'}"/>
 
-<semui:modal id="${modalID ?: 'copyFilteredEmailAddresses_ajaxModal'}" text="${message(code:'menu.institutions.copy_emailaddresses', args:[orgList?.size()?:0])}" hideSubmitButton="true">
+<semui:modal id="${modalID ?: 'copyFilteredEmailAddresses_ajaxModal'}" text="${orgList ? message(code:'menu.institutions.copy_emailaddresses', args:[orgList?.size()?:0]) : message(code:'menu.institutions.copy_emailaddresses.button')}" hideSubmitButton="true">
```
Comment (klodav) [inline app/grails-app/views/templates/_copyFilteredEmailAddresses.gsp:7]: orgList.size()

File: app/grails-app/views/organisation/show.gsp [hunk 11]
```diff
@@ -310,223 +310,98 @@
...
+                                               data-semui="modal"
+                                               data-href="#personFormModalTechnichalSupport"/>
 
-                            </dt>
-                            <dd>
+                                        <g:render template="/person/formModal"
+                                                  model="[tenant            : contextOrg,
+                                                          org               : orgInstance,
+                                                          isPublic          : true,
+                                                          presetFunctionType: RefdataValue.getByValueAndCategory('Technichal Support', RDConstants.PERSON_FUNCTION),
```
Comment (klodav) [inline app/grails-app/views/organisation/show.gsp:348]: RDStore.PRS_FUNC_TECHNICAL_SUPPORT

File: app/grails-app/views/organisation/show.gsp [hunk 12]
```diff
@@ -310,223 +310,98 @@
...
-                                        orgInstance.addresses.each{
-                                            String typeName = it.type?.getI10n('value')
+                                <%
+                                    Set<String> typeNames = new TreeSet<String>()
+                                    typeNames.add(RDStore.ADRESS_TYPE_BILLING.getI10n('value'))
+                                    typeNames.add(RDStore.ADRESS_TYPE_POSTAL.getI10n('value'))
+                                    Map<String, List> typeAddressMap = [:]
+                                    orgInstance.addresses.each {
+                                        it.type.each { type ->
+                                            String typeName = type?.getI10n('value')
```
Comment (klodav) [inline app/grails-app/views/organisation/show.gsp:365]: type?

File: app/grails-app/views/organisation/addressbook.gsp [hunk 13]
```diff
@@ -2,50 +2,145 @@
...
-		<g:if test="${visiblePersons}">
-			<g:render template="/templates/cpa/person_table" model="${[
-                    persons: visiblePersons,
-                    restrictToOrg: orgInstance,
-                    tmplConfigShow: ['lineNumber', 'organisation', 'functionPosition', 'name', 'showContacts', 'showAddresses']
-            ]}" />
-		</g:if>
+<g:javascript>
+    function personCreate(contactFor, org) {
+        var url = '<g:createLink controller="ajax"
```
Comment (klodav) [inline app/grails-app/views/organisation/addressbook.gsp:124]: controller="ajaxHtml"

Comment (klodav) [reply app/grails-app/views/organisation/addressbook.gsp:124]: '&showAddresses=false&showContacts=true'

File: app/grails-app/views/templates/_copyFilteredEmailAddresses.gsp (hunk 10 above)
Comment (Moetez16) [reply app/grails-app/views/templates/_copyFilteredEmailAddresses.gsp:7]: orgList kann auch nicht vergeben sein!

**Comment by Moetez16** [comment]
Done

**Review by klodav — APPROVED** [review]

**Pull request merged**